### ISSUE TITLE
Defective bp4 dataset with v2.6.0, works with v2.5.0

### ISSUE THREAD

**Issue by williamfgc** [ornladios/ADIOS2#2420]
**Describe the bug**
BP4 dataset generated using the same code  

**To Reproduce**
Attached a good file and a bad file in `adios2_bug_data.tar.gz`
good file (with 2.5.0): ex9-p-periodic-hexagon.mesh_o3_250.bp
bad file (with 2.6.0): ex9-p-periodic-hexagon.mesh_o3_260_bad.bp
bp4dgb output for bad file: bp4dgb.log with message `ERROR: Var Name string length (8) is longer than the limit to stay inside the block (-12.0)`
[adios2_bug_data.tar.gz](https://github.com/ornladios/ADIOS2/files/5059696/adios2_bug_data.tar.gz)

**Expected behavior**
File contents should be the same and correct

**Desktop (please complete the following information):**
 - OS/Platform: Ubuntu 18.04
 - gcc 8

**Additional context**
Files generated from MFEM-adios2 integration in ExAM using exactly the same source code for adios2 v2.5.0 and v2.6.0 this could affect other apps since these were working fine with v2.5.0 standard production features.

**Comment by williamfgc** [comment]
@pnorbert FYI, this has been stable with v2.5.0 all across.

**Comment by pnorbert** [comment]
The md.idx of the 260 version is already corrupt. The offsets into the md.0 file should be monotonous and completely filling the space (EndPtr of step N = PGPtr of step N+1) but here record 6 is not continuing record 5 and later you can see many jumps backwards. Was this a single run writing this file? I don't have a clue how it is possible to create such an index. 
```
$ bp4dbg ex9-p-periodic-hexagon.mesh_o3_260_bad.bp/md.idx 
========================================================
    Index Table File: ex9-p-periodic-hexagon.mesh_o3_260_bad.bp/md.idx
========================================================
----------------------------------------------------------------------------------------------------------------------
|        Version string            | Major | Minor | Patch | unused | Endian | BP version | Active |     unused      |
|          32 bytes                |   1B  |   1B  |   1B  |   1B   |   1B   |     1B     |   1B   |       25B       |
+--------------------------------------------------------------------------------------------------------------------+
| ADIOS-BP v2.6.0 Index Table      |   2   |   6   |   0   |        |  yes   |      4     |   no   |                 |
----------------------------------------------------------------------------------------------------------------------
 
-----------------------------------------------------------------------------------------------------------
|   Step   |   Rank   |    PGPtr    |   VarPtr    |   AttPtr    |   EndPtr    |  Timestamp  |    unused   |
+---------------------------------------------------------------------------------------------------------+
|        1 |        0 |          64 |         137 |         911 |        2290 |  1597182928 |           0 |
|        2 |        0 |        2501 |        2574 |        2840 |        2852 |  1597182928 |           0 |
|        3 |        0 |        2852 |        2925 |        3191 |        3203 |  1597182928 |           0 |
|        4 |        0 |        3203 |        3276 |        3542 |        3554 |  1597182928 |           0 |
|        5 |        0 |        3554 |        3627 |        3893 |        3905 |  1597182928 |           0 |
|        6 |        0 |        3122 |        3195 |        3318 |        3330 |  1597182928 |           0 |
|        7 |        0 |        3330 |        3403 |        3526 |        3538 |  1597182928 |           0 |
|        8 |        0 |        3538 |        3611 |        3734 |        3746 |  1597182928 |           0 |
|        9 |        0 |        3746 |        3819 |        3942 |        3954 |  1597182928 |           0 |
|       10 |        0 |        3954 |        4028 |        4151 |        4163 |  1597182928 |           0 |
|       11 |        0 |        4163 |        4237 |        4360 |        4372 |  1597182928 |           0 |
|       12 |        0 |        4372 |        4446 |        4569 |        4581 |  1597182928 |           0 |
|       13 |        0 |        6365 |        6439 |        6705 |        6717 |  1597182928 |           0 |
|       14 |        0 |        6717 |        6791 |        7057 |        7069 |  1597182928 |           0 |
|       15 |        0 |        7069 |        7143 |        7409 |        7421 |  1597182928 |           0 |
|       16 |        0 |        7421 |        7495 |        7761 |        7773 |  1597182928 |           0 |
|       17 |        0 |        7773 |        7847 |        8113 |        8125 |  1597182928 |           0 |
|       18 |        0 |        5626 |        5700 |        5823 |        5835 |  1597182928 |           0 |
|       19 |        0 |        5835 |        5909 |        6032 |        6044 |  1597182928 |           0 |
|       20 |        0 |        6044 |        6118 |        6241 |        6253 |  1597182928 |           0 |
|       21 |        0 |        9181 |        9255 |        9521 |        9533 |  1597182928 |           0 |
|       22 |        0 |        9533 |        9607 |        9873 |        9885 |  1597182928 |           0 |
|       23 |        0 |        9885 |        9959 |       10225 |       10237 |  1597182928 |           0 |
|       24 |        0 |       10237 |       10311 |       10577 |       10589 |  1597182928 |           0 |
|       25 |        0 |       10589 |       10663 |       10929 |       10941 |  1597182928 |           0 |
|       26 |        0 |       10941 |       11015 |       11281 |       11293 |  1597182928 |           0 |
|       27 |        0 |       11293 |       11367 |       11633 |       11645 |  1597182928 |           0 |
|       28 |        0 |        7716 |        7790 |        7913 |        7925 |  1597182928 |           0 |
|       29 |        0 |        7925 |        7999 |        8122 |        8134 |  1597182928 |           0 |
|       30 |        0 |        8134 |        8208 |        8331 |        8343 |  1597182928 |           0 |
|       31 |        0 |        8343 |        8417 |        8540 |        8552 |  1597182928 |           0 |
|       32 |        0 |        8552 |        8626 |        8749 |        8761 |  1597182928 |           0 |
|       33 |        0 |       13405 |       13479 |       13745 |       13757 |  1597182928 |           0 |
|       34 |        0 |        8970 |        9044 |        9167 |        9179 |  1597182928 |           0 |
|       35 |        0 |        9179 |        9253 |        9376 |        9388 |  1597182928 |           0 |
|       36 |        0 |        9388 |        9462 |        9585 |        9597 |  1597182928 |           0 |
|       37 |        0 |       14813 |       14887 |       15153 |       15165 |  1597182928 |           0 |
|       38 |        0 |       15165 |       15239 |       15505 |       15517 |  1597182928 |           0 |
|       39 |        0 |       15517 |       15591 |       15857 |       15869 |  1597182928 |           0 |
|       40 |        0 |       15869 |       15943 |       16209 |       16221 |  1597182928 |           0 |
|       41 |        0 |       16221 |       16295 |       16561 |       16573 |  1597182928 |           0 |
|       42 |        0 |       16573 |       16647 |       16913 |       16925 |  1597182928 |           0 |
|       43 |        0 |       10851 |       10925 |       11048 |       11060 |  1597182928 |           0 |
|       44 |        0 |       17277 |       17351 |       17617 |       17629 |  1597182928 |           0 |
|       45 |        0 |       11269 |       11343 |       11466 |       11478 |  1597182928 |           0 |
|       46 |        0 |       17981 |       18055 |       18321 |       18333 |  1597182928 |           0 |
|       47 |        0 |       18333 |       18407 |       18673 |       18685 |  1597182928 |           0 |
|       48 |        0 |       18685 |       18759 |       19025 |       19037 |  1597182928 |           0 |
|       49 |        0 |       19037 |       19111 |       19377 |       19389 |  1597182928 |           0 |
|       50 |        0 |       19389 |       19463 |       19729 |       19741 |  1597182928 |           0 |
|       51 |        0 |       19741 |       19815 |       20081 |       20093 |  1597182928 |           0 |
|       52 |        0 |       20093 |       20167 |       20433 |       20445 |  1597182928 |           0 |
|       53 |        0 |       20445 |       20519 |       20785 |       20797 |  1597182928 |           0 |
|       54 |        0 |       20797 |       20871 |       21137 |       21149 |  1597182928 |           0 |
|       55 |        0 |       21149 |       21223 |       21489 |       21501 |  1597182928 |           0 |
|       56 |        0 |       13568 |       13642 |       13765 |       13777 |  1597182928 |           0 |
|       57 |        0 |       21853 |       21927 |       22193 |       22205 |  1597182928 |           0 |
|       58 |        0 |       22205 |       22279 |       22545 |       22557 |  1597182928 |           0 |
|       59 |        0 |       22557 |       22631 |       22897 |       22909 |  1597182928 |           0 |
|       60 |        0 |       22909 |       22983 |       23249 |       23261 |  1597182928 |           0 |
|       61 |        0 |       23261 |       23335 |       23601 |       23613 |  1597182928 |           0 |
|       62 |        0 |       23613 |       23687 |       23953 |       23965 |  1597182928 |           0 |
|       63 |        0 |       15031 |       15105 |       15228 |       15240 |  1597182928 |           0 |
|       64 |        0 |       24317 |       24391 |       24657 |       24669 |  1597182928 |           0 |
|       65 |        0 |       24669 |       24743 |       25009 |       25021 |  1597182928 |           0 |
|       66 |        0 |       15658 |       15732 |       15855 |       15867 |  1597182928 |           0 |
|       67 |        0 |       15867 |       15941 |       16064 |       16076 |  1597182928 |           0 |
|       68 |        0 |       16076 |       16150 |       16273 |       16285 |  1597182928 |           0 |
|       69 |        0 |       16285 |       16359 |       16482 |       16494 |  1597182928 |           0 |
|       70 |        0 |       26429 |       26503 |       26769 |       26781 |  1597182928 |           0 |
|       71 |        0 |       16703 |       16777 |       16900 |       16912 |  1597182928 |           0 |
|       72 |        0 |       27133 |       27207 |       27473 |       27485 |  1597182928 |           0 |
|       73 |        0 |       27485 |       27559 |       27825 |       27837 |  1597182928 |           0 |
|       74 |        0 |       17330 |       17404 |       17527 |       17539 |  1597182928 |           0 |
|       75 |        0 |       28189 |       28263 |       28529 |       28541 |  1597182928 |           0 |
|       76 |        0 |       28541 |       28615 |       28881 |       28893 |  1597182928 |           0 |
|       77 |        0 |       28893 |       28967 |       29233 |       29245 |  1597182928 |           0 |
|       78 |        0 |       29245 |       29319 |       29585 |       29597 |  1597182928 |           0 |
|       79 |        0 |       29597 |       29671 |       29937 |       29949 |  1597182928 |           0 |
|       80 |        0 |       18584 |       18658 |       18781 |       18793 |  1597182928 |           0 |
|       81 |        0 |       30301 |       30375 |       30641 |       30653 |  1597182928 |           0 |
|       82 |        0 |       30653 |       30727 |       30993 |       31005 |  1597182928 |           0 |
|       83 |        0 |       31005 |       31079 |       31345 |       31357 |  1597182928 |           0 |
|       84 |        0 |       31357 |       31431 |       31697 |       31709 |  1597182928 |           0 |
|       85 |        0 |       31709 |       31783 |       32049 |       32061 |  1597182928 |           0 |
|       86 |        0 |       32061 |       32135 |       32401 |       32413 |  1597182928 |           0 |
|       87 |        0 |       32413 |       32487 |       32753 |       32765 |  1597182929 |           0 |
|       88 |        0 |       32765 |       32839 |       33105 |       33117 |  1597182929 |           0 |
|       89 |        0 |       33117 |       33191 |       33457 |       33469 |  1597182929 |           0 |
|       90 |        0 |       33469 |       33543 |       33809 |       33821 |  1597182929 |           0 |
|       91 |        0 |       33821 |       33895 |       34161 |       34173 |  1597182929 |           0 |
|       92 |        0 |       34173 |       34247 |       34513 |       34525 |  1597182929 |           0 |
|       93 |        0 |       34525 |       34599 |       34865 |       34877 |  1597182929 |           0 |
|       94 |        0 |       34877 |       34951 |       35217 |       35229 |  1597182929 |           0 |
|       95 |        0 |       35229 |       35303 |       35569 |       35581 |  1597182929 |           0 |
|       96 |        0 |       35581 |       35655 |       35921 |       35933 |  1597182929 |           0 |
|       97 |        0 |       35933 |       36007 |       36273 |       36285 |  1597182929 |           0 |
|       98 |        0 |       36285 |       36359 |       36625 |       36637 |  1597182929 |           0 |
|       99 |        0 |       36637 |       36711 |       36977 |       36989 |  1597182929 |           0 |
|      100 |        0 |       22764 |       22839 |       22962 |       22974 |  1597182929 |           0 |
|      101 |        0 |       37342 |       37417 |       37683 |       37695 |  1597182929 |           0 |
|      102 |        0 |       37695 |       37770 |       38036 |       38048 |  1597182929 |           0 |
|      103 |        0 |       23394 |       23469 |       23592 |       23604 |  1597182929 |           0 |
|      104 |        0 |       38401 |       38476 |       38742 |       38754 |  1597182929 |           0 |
|      105 |        0 |       23814 |       23889 |       24012 |       24024 |  1597182929 |           0 |
|      106 |        0 |       39107 |       39182 |       39448 |       39460 |  1597182929 |           0 |
|      107 |        0 |       24234 |       24309 |       24432 |       24444 |  1597182929 |           0 |
|      108 |        0 |       39813 |       39888 |       40154 |       40166 |  1597182929 |           0 |
|      109 |        0 |       24654 |       24729 |       24852 |       24864 |  1597182929 |           0 |
|      110 |        0 |       40519 |       40594 |       40860 |       40872 |  1597182929 |           0 |
|      111 |        0 |       40872 |       40947 |       41213 |       41225 |  1597182929 |           0 |
|      112 |        0 |       41225 |       41300 |       41566 |       41578 |  1597182929 |           0 |
|      113 |        0 |       41578 |       41653 |       41919 |       41931 |  1597182929 |           0 |
|      114 |        0 |       41931 |       42006 |       42272 |       42284 |  1597182929 |           0 |
|      115 |        0 |       42284 |       42359 |       42625 |       42637 |  1597182929 |           0 |
|      116 |        0 |       42637 |       42712 |       42978 |       42990 |  1597182929 |           0 |
|      117 |        0 |       26334 |       26409 |       26532 |       26544 |  1597182929 |           0 |
|      118 |        0 |       43343 |       43418 |       43684 |       43696 |  1597182929 |           0 |
|      119 |        0 |       43696 |       43771 |       44037 |       44049 |  1597182929 |           0 |
|      120 |        0 |       44049 |       44124 |       44390 |       44402 |  1597182929 |           0 |
|      121 |        0 |       44402 |       44477 |       44743 |       44755 |  1597182929 |           0 |
|      122 |        0 |       44755 |       44830 |       45096 |       45108 |  1597182929 |           0 |
|      123 |        0 |       45108 |       45183 |       45449 |       45461 |  1597182929 |           0 |
|      124 |        0 |       45461 |       45536 |       45802 |       45814 |  1597182929 |           0 |
|      125 |        0 |       45814 |       45889 |       46155 |       46167 |  1597182929 |           0 |
|      126 |        0 |       46167 |       46242 |       46508 |       46520 |  1597182929 |           0 |
|      127 |        0 |       46520 |       46595 |       46861 |       46873 |  1597182929 |           0 |
|      128 |        0 |       46873 |       46948 |       47214 |       47226 |  1597182929 |           0 |
|      129 |        0 |       28854 |       28929 |       29052 |       29064 |  1597182929 |           0 |
|      130 |        0 |       29064 |       29139 |       29262 |       29274 |  1597182929 |           0 |
|      131 |        0 |       47932 |       48007 |       48273 |       48285 |  1597182929 |           0 |
|      132 |        0 |       48285 |       48360 |       48626 |       48638 |  1597182929 |           0 |
|      133 |        0 |       29694 |       29769 |       29892 |       29904 |  1597182929 |           0 |
|      134 |        0 |       48991 |       49066 |       49332 |       49344 |  1597182929 |           0 |
|      135 |        0 |       30114 |       30189 |       30312 |       30324 |  1597182929 |           0 |
|      136 |        0 |       30324 |       30399 |       30522 |       30534 |  1597182929 |           0 |
|      137 |        0 |       50050 |       50125 |       50391 |       50403 |  1597182929 |           0 |
|      138 |        0 |       50403 |       50478 |       50744 |       50756 |  1597182929 |           0 |
|      139 |        0 |       50756 |       50831 |       51097 |       51109 |  1597182929 |           0 |
|      140 |        0 |       51109 |       51184 |       51450 |       51462 |  1597182929 |           0 |
|      141 |        0 |       51462 |       51537 |       51803 |       51815 |  1597182929 |           0 |
|      142 |        0 |       51815 |       51890 |       52156 |       52168 |  1597182929 |           0 |
|      143 |        0 |       52168 |       52243 |       52509 |       52521 |  1597182929 |           0 |
|      144 |        0 |       52521 |       52596 |       52862 |       52874 |  1597182929 |           0 |
|      145 |        0 |       52874 |       52949 |       53215 |       53227 |  1597182929 |           0 |
|      146 |        0 |       53227 |       53302 |       53568 |       53580 |  1597182929 |           0 |
|      147 |        0 |       53580 |       53655 |       53921 |       53933 |  1597182929 |           0 |
|      148 |        0 |       53933 |       54008 |       54274 |       54286 |  1597182929 |           0 |
|      149 |        0 |       33054 |       33129 |       33252 |       33264 |  1597182929 |           0 |
|      150 |        0 |       54639 |       54714 |       54980 |       54992 |  1597182929 |           0 |
|      151 |        0 |       54992 |       55067 |       55333 |       55345 |  1597182929 |           0 |
|      152 |        0 |       55345 |       55420 |       55686 |       55698 |  1597182929 |           0 |
|      153 |        0 |       55698 |       55773 |       56039 |       56051 |  1597182929 |           0 |
|      154 |        0 |       56051 |       56126 |       56392 |       56404 |  1597182929 |           0 |
|      155 |        0 |       56404 |       56479 |       56745 |       56757 |  1597182929 |           0 |
|      156 |        0 |       56757 |       56832 |       57098 |       57110 |  1597182929 |           0 |
|      157 |        0 |       57110 |       57185 |       57451 |       57463 |  1597182929 |           0 |
|      158 |        0 |       57463 |       57538 |       57804 |       57816 |  1597182929 |           0 |
|      159 |        0 |       57816 |       57891 |       58157 |       58169 |  1597182929 |           0 |
|      160 |        0 |       58169 |       58244 |       58510 |       58522 |  1597182929 |           0 |
|      161 |        0 |       58522 |       58597 |       58863 |       58875 |  1597182929 |           0 |
|      162 |        0 |       58875 |       58950 |       59216 |       59228 |  1597182929 |           0 |
|      163 |        0 |       59228 |       59303 |       59569 |       59581 |  1597182929 |           0 |
|      164 |        0 |       36204 |       36279 |       36402 |       36414 |  1597182929 |           0 |
|      165 |        0 |       59934 |       60009 |       60275 |       60287 |  1597182929 |           0 |
|      166 |        0 |       60287 |       60362 |       60628 |       60640 |  1597182929 |           0 |
|      167 |        0 |       60640 |       60715 |       60981 |       60993 |  1597182929 |           0 |
|      168 |        0 |       60993 |       61068 |       61334 |       61346 |  1597182929 |           0 |
|      169 |        0 |       37254 |       37329 |       37452 |       37464 |  1597182929 |           0 |
|      170 |        0 |       61699 |       61774 |       62040 |       62052 |  1597182929 |           0 |
|      171 |        0 |       62052 |       62127 |       62393 |       62405 |  1597182929 |           0 |
|      172 |        0 |       37884 |       37959 |       38082 |       38094 |  1597182929 |           0 |
|      173 |        0 |       38094 |       38169 |       38292 |       38304 |  1597182929 |           0 |
|      174 |        0 |       63111 |       63186 |       63452 |       63464 |  1597182929 |           0 |
|      175 |        0 |       63464 |       63539 |       63805 |       63817 |  1597182929 |           0 |
|      176 |        0 |       63817 |       63892 |       64158 |       64170 |  1597182929 |           0 |
|      177 |        0 |       64170 |       64245 |       64511 |       64523 |  1597182929 |           0 |
|      178 |        0 |       64523 |       64598 |       64864 |       64876 |  1597182929 |           0 |
|      179 |        0 |       64876 |       64951 |       65217 |       65229 |  1597182929 |           0 |
|      180 |        0 |       65229 |       65304 |       65570 |       65582 |  1597182929 |           0 |
|      181 |        0 |       65582 |       65657 |       65923 |       65935 |  1597182929 |           0 |
|      182 |        0 |       65935 |       66010 |       66276 |       66288 |  1597182929 |           0 |
|      183 |        0 |       66288 |       66363 |       66629 |       66641 |  1597182929 |           0 |
|      184 |        0 |       66641 |       66716 |       66982 |       66994 |  1597182929 |           0 |
|      185 |        0 |       66994 |       67069 |       67335 |       67347 |  1597182929 |           0 |
|      186 |        0 |       67347 |       67422 |       67688 |       67700 |  1597182929 |           0 |
|      187 |        0 |       67700 |       67775 |       68041 |       68053 |  1597182929 |           0 |
|      188 |        0 |       41244 |       41319 |       41442 |       41454 |  1597182929 |           0 |
|      189 |        0 |       68406 |       68481 |       68747 |       68759 |  1597182929 |           0 |
|      190 |        0 |       68759 |       68834 |       69100 |       69112 |  1597182929 |           0 |
|      191 |        0 |       41874 |       41949 |       42072 |       42084 |  1597182929 |           0 |
|      192 |        0 |       69465 |       69540 |       69806 |       69818 |  1597182929 |           0 |
|      193 |        0 |       42294 |       42369 |       42492 |       42504 |  1597182929 |           0 |
|      194 |        0 |       70171 |       70246 |       70512 |       70524 |  1597182929 |           0 |
|      195 |        0 |       70524 |       70599 |       70865 |       70877 |  1597182929 |           0 |
|      196 |        0 |       70877 |       70952 |       71218 |       71230 |  1597182929 |           0 |
|      197 |        0 |       71230 |       71305 |       71571 |       71583 |  1597182929 |           0 |
|      198 |        0 |       71583 |       71658 |       71924 |       71936 |  1597182929 |           0 |
|      199 |        0 |       71936 |       72011 |       72277 |       72289 |  1597182929 |           0 |
|      200 |        0 |       72289 |       72364 |       72630 |       72642 |  1597182929 |           0 |
|      201 |        0 |       72642 |       72717 |       72983 |       72995 |  1597182929 |           0 |
-----------------------------------------------------------------------------------------------------------
```

**Comment by pnorbert** [comment]
These were not the same runs just with different versions of ADIOS2, were they? I see 4 blocks in v250 and 1 block in v260 for each variable. Also, the v250 has CYCLE and TIME in it, but v260 doesn't (in the first step). Why is that?

I trimmed the index to see only the first step. 
```
mv md.idx md-full.idx
dd if=md-full.idx of=md.idx bs=64 count=2` 
```

```
$ bpls -l x250.bp/
  int32_t   CYCLE          scalar = 0
  uint32_t  NumOfElements  {4} = 47 / 49
  uint32_t  NumOfVertices  {4} = 188 / 196
  double    TIME           scalar = 0
  uint64_t  connectivity   [4]*{__, 5} = 0 / 195
  int32_t   material       [4]*{__} = 1 / 3
  double    solution       [4]*{__} = 4.51835e-42 / 0.99765
  uint32_t  types          scalar = 9
  double    vertices       [4]*{__, 2} = -1 / 1

$ bpls -l x260.bp/
  uint32_t  NumOfElements  {1} = 49 / 49
  uint32_t  NumOfVertices  {1} = 196 / 196
  uint64_t  connectivity   [1]*{49, 5} = 0 / 195
  int32_t   material       [1]*{49} = 1 / 3
  double    solution       [1]*{196} = 5.2278e-33 / 0.995317
  uint32_t  types          scalar = 9
  double    vertices       [1]*{196, 2} = -0.866025 / 1
```

TIME is also missing from the VTK Schema

```
$ bpls -lA x250.bp vtk.xml
  string    vtk.xml                      attr   = 
<VTKFile type="UnstructuredGrid" version="0.2" byte_order="LittleEndian">
  <UnstructuredGrid>
    <Piece NumberOfPoints="NumOfVertices" NumberOfCells="NumOfElements">
      <Points>
        <DataArray Name="vertices" />
      </Points>
      <CellData>
        <DataArray Name="material" />
      </CellData>
      <Cells>
        <DataArray Name="connectivity" />
        <DataArray Name="types" />
      </Cells>
      <PointData>
        <DataArray Name="solution" />
        <DataArray Name="TIME">
          TIME
        </DataArray>
      </PointData>
    </Piece>
  </UnstructuredGrid>
</VTKFile>

$ bpls -lA x260.bp vtk.xml
  string    vtk.xml                      attr   = 
<VTKFile type="UnstructuredGrid" version="0.2" byte_order="LittleEndian">
  <UnstructuredGrid>
    <Piece NumberOfPoints="NumOfVertices" NumberOfCells="NumOfElements">
      <Points>
        <DataArray Name="vertices" />
      </Points>
      <CellData>
        <DataArray Name="material" />
      </CellData>
      <Cells>
        <DataArray Name="connectivity" />
        <DataArray Name="types" />
      </Cells>
      <PointData>
        <DataArray Name="solution" />
      </PointData>
    </Piece>
  </UnstructuredGrid>
</VTKFile>
```

**Comment by williamfgc** [comment]
> These were not the same runs just with different versions of ADIOS2

Correct, it is the same source code linking to two adios2 versions. @pnorbert thanks for looking into, helps narrowing it down to to how 2.6.0 new library stack is linked to MFEM as it uses legacy CMake. It might help if the bp4dbg error is caught early in bp4 as it silently writes the invalid bp4 dataset. Are you using `bpls` from master? I can't get bpls to work on the defective dataset. Thanks.

**Comment by pnorbert** [comment]
I am trimming the md.idx to (N+1)*64 bytes to see only N steps. bpls works for 1 step of v260. The second step's offsets in the md.idx are already invalid, so the metadata parser reads garbage. See my first post: step 2 PGPtr should be 2290 but it is 2501, which points into a random place in step 4 actually. 

So the v2.6 run used a single process run? Or was this a parallel run? The big change from 2.5 to 2.6 was the rework of the serial/MPI library and the use of adiosComm class. I wonder if your parallel run somehow used the serial library and multiple processes were messing up the file content.

**Comment by williamfgc** [comment]
> I wonder if your parallel run somehow used the serial library and multiple processes were messing up the file content.

Yes, it is a parallel run, that seems to be the most likely scenario from the `bpls` output you are showing. Depending on how I link, `libadios2*mpi*.so` might not be picked up automatically with the same setup for v2.5.0 using MPI_Comm constructors. I'll run the debugger to see if we are using MPI at all. Thanks.

**Comment by williamfgc** [comment]
@pnorbert I got this working, thanks for your help. The definition `-DADIOS2_USE_MPI` must be added to upstream compilation for the `libadios2_cxx11_mpi.so.2` library to be linked, otherwise we run on serial mode always. Essentially the valid MPI constructor is ignored and demotes to a serial constructor. 

Providing my feedback to @bradking as the above could be an error with a nice message as well:

```
> ldd exe
libadios2_cxx11.so.2 => /opt/adios2/2.6.0/lib/libadios2_cxx11.so.2 
```

```
adios2::ADIOS adios(comm);   // throws "tried using a valid comm, but missing -DADIOS2_USE_MPI definition in upstream compilation
```

After the fix:
```
> ldd exe
libadios2_cxx11_mpi.so.2 => /opt/adios2/2.6.0/lib/libadios2_cxx11_mpi.so.2 
libadios2_cxx11.so.2 => /opt/adios2/2.6.0/lib/libadios2_cxx11.so.2 
```

The above can be easily reproduced with legacy non-target based cmake (this is out of my control). Another option is to document or provide a `FindADIOS2.cmake` for legacy non-target based cmake (which is still very much alive).

**Comment by bradking** [comment]
> Essentially the valid MPI constructor is ignored and demotes to a serial constructor.
> ```c++
> adios2::ADIOS adios(comm);
> ```

What is the type of `comm` in that example?  Without `-DADIOS2_USE_MPI` the MPI constructor is [removed](https://github.com/ornladios/ADIOS2/blob/v2.6.0/bindings/CXX11/adios2/cxx11/ADIOS.h#L43-L65) via preprocessor conditions.  Calling it should be a compile-time error.

**Comment by williamfgc** [comment]
@bradking thanks for chiming in. To reproduce removing [this line](https://github.com/mfem/mfem/blob/adios2-element-attribute/config/cmake/modules/FindADIOS2.cmake#L51) in MFEM's `FindADIOS2.cmake` (it's currently a PR to be reviewed, use this branch rather master) will have the effect. The constructor for the examples is `MPI_COMM_WORLD`, I'm using [ex9p](https://github.com/mfem/mfem/blob/adios2-element-attribute/examples/ex9p.cpp) as my baseline. The ADIOS2 DataCollection starts [here](https://github.com/mfem/mfem/blob/adios2-element-attribute/examples/ex9p.cpp#L435). Let me know if you have any questions, happy to find out what's going on.

**Comment by bradking** [comment]
Thanks.  With that I've been able to identify the problem.  The code on [this line](https://github.com/mfem/mfem/blob/00550ef4f737a29208bcc9b1be6d200723cd6882/general/adios2stream.cpp#L118) was expected to cause a compiler error without `-DADIOS2_USE_MPI`.  However, it happens that `MPI_Comm` is typically implemented either as a pointer type or an integer type.  Either of those can be converted to `bool` implicitly, so the serial [debug mode constructor](https://github.com/ornladios/ADIOS2/blob/v2.6.0/bindings/CXX11/adios2/cxx11/ADIOS.h#L76-L81) ends up getting selected.

I've opened #2434 with an attempt to make this case an error for clients.

**Comment by williamfgc** [comment]
@bradking Great, thanks for fixing this! Makes sense. I'm closing this, great work!

**Comment by williamfgc** [comment]
Closed via #2434
